### ISSUE TITLE
Parallelization of tf.DistributedDataset Prefetch? 

### ISSUE THREAD

**Issue by levimcclenny** [tensorflow/tensorflow#44820]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04 - DGX OS 4
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Pip
- TensorFlow version (use command below): 2.3.1
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.1
- GPU model and memory: 4x V100 32GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
DistributedDataset doesnt seem to keep up with input - CPU utilization is low but docs only provide information for tf.Dataset, not tf.DistributedDataset. Curious if there are ways to use AUTOTUNE on a DiatributedDataset to parallelize the prefetching operations. 

**Describe the expected behavior**
Prefetch allows for the data to come off host RAM quick enough to be utilized rapidly.

**Standalone code to reproduce the issue**

This is part of a package I'm developing, so standalone code is rather difficult. Below is the fit function for the model to be trained. Unfortunately it wont be possible to reproduce it standalone. 

obj is a class that is being passed in. This contains the loss, etc. Really the question lies in the first ~30 lines where the dataset is turned into a tf.distributeddataset and then in the last bit where it is iterated

```
def fit_dist(obj, tf_iter, newton_iter, batch_sz = None, newton_eager = True):

    BUFFER_SIZE = len(obj.x_f)
    EPOCHS = tf_iter
    obj.strategy = tf.distribute.MirroredStrategy(cross_device_ops = tf.distribute.NcclAllReduce())
    print("number of devices: {}".format(obj.strategy.num_replicas_in_sync))

    if batch_sz is not None:
        obj.batch_sz = batch_sz
    else:
        obj.batch_sz = len(obj.x_f)

    N_f = len(obj.x_f)
    n_batches =  N_f // obj.batch_sz

    BATCH_SIZE_PER_REPLICA = obj.batch_sz
    GLOBAL_BATCH_SIZE = BATCH_SIZE_PER_REPLICA * obj.strategy.num_replicas_in_sync

    train_dataset = tf.data.Dataset.from_tensor_slices((obj.x_f, obj.t_f)).batch(GLOBAL_BATCH_SIZE)

    #print(GLOBAL_BATCH_SIZE)
    obj.train_dist_dataset = obj.strategy.experimental_distribute_dataset(train_dataset)


    start_time = time.time()

    with obj.strategy.scope():
        obj.u_model = neural_net(obj.layer_sizes)
        obj.tf_optimizer = tf.keras.optimizers.Adam(lr = 0.005, beta_1=.99)
        obj.tf_optimizer_weights = tf.keras.optimizers.Adam(lr = 0.005, beta_1=.99)
        #Can adjust batch size for collocation points, here we set it to N_f

        if obj.isAdaptive:
            obj.col_weights = tf.Variable(tf.random.uniform([50000, 1]))
            obj.u_weights = tf.Variable(100*tf.random.uniform([200, 1]))


    print("starting Adam training")
    STEPS = np.max((n_batches // obj.strategy.num_replicas_in_sync,1))
    
    for epoch in range(tf_iter):
            train_loss = train_epoch(obj, obj.train_dist_dataset, obj.col_weights, STEPS)
            if epoch == 2:
                tf.profiler.experimental.start('../cache/tblogdir1')
            if epoch % 10 == 0:
                elapsed = time.time() - start_time
                print('It: %d, Time: %.2f' % (epoch, elapsed))
                tf.print(f"total loss: {train_loss}")
                start_time = time.time()

    tf.profiler.experimental.stop()
    #l-bfgs-b optimization
    print("Starting L-BFGS training")
    lbfgs_train(obj, newton_iter)

@tf.function
def train_epoch(obj, dataset, col_weights, STEPS):
    total_loss = 0.0
    num_batches = 0.0
    #dist_col_weights = iter(col_weights)
    dist_dataset_iterator = iter(dataset)
    for _ in range(STEPS):
        total_loss += distributed_train_step(obj, next(dist_dataset_iterator), col_weights)
        num_batches += 1
    train_loss = total_loss / num_batches
    return train_loss

def train_step(obj, inputs, col_weights):
    obj.dist_x_f, obj.dist_t_f = inputs
    obj.dist_col_weights = col_weights
    if obj.isAdaptive:
        obj.variables = obj.u_model.trainable_variables
        obj.variables.extend([obj.u_weights, obj.dist_col_weights])
        loss_value, mse_0, mse_b, mse_f, grads = obj.grad()
        obj.tf_optimizer.apply_gradients(zip(grads[:-2], obj.u_model.trainable_variables))
        obj.tf_optimizer_weights.apply_gradients(zip([-grads[-2], -grads[-1]], [obj.u_weights, obj.dist_col_weights]))
    else:
        obj.variables = obj.u_model.trainable_variables
        loss_value, mse_0, mse_b, mse_f, grads = obj.grad()
        obj.tf_optimizer.apply_gradients(zip(grads, obj.u_model.trainable_variables))
    return loss_value

def distributed_train_step(obj, dataset_inputs, col_weights):
    per_replica_losses = obj.strategy.run(train_step, args=(obj, dataset_inputs, col_weights))
    return obj.strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_losses,
                         axis=None)
```

**What I have tried**
I have been through the docs in custom training loops with mirrored strategy, distributed dataset training, distributed training with Keras, as well as optimizing on GPU with the profiler, better performance with tf.data API, and Optimize TensorFlow performance using the Profiler, most of the associated how-tos and self helps.

Really my question is about more support with tf.DistributedDataset - It prefetches alright (I checked the traces) but it doesn't really utilize the CPU fully (it would appear) and the ops on GPU outrun the data fetching. There is no data preprocessing for my training, all it has to do is pull the points (literally coordinates, a tuple of 2 tf.float32 values) from memory. They're not images or anything large. This is why I think the GPU processing is simply outrunning the data fetching, but I wanted to see if there was a way to better parallelize the tf.DistributedDataset memory fetches on CPU, etc to help get more data in the GPU faster. Like is there an "autotune" that I need to enable to do that better in tf.distributedDataset?

I'm attaching my profile traces. Another question would be why only 9% of my ops live on the GPU. When I run on a single GPU without mirrored strategy all the ops live on GPU, here only 9% do. These two event are likely related, but I cant seem to coerce tf into putting more ops on GPU. 

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

[2020_11_13_04_12_22.zip](https://github.com/tensorflow/tensorflow/files/5534766/2020_11_13_04_12_22.zip)

**Comment by nikitamaia** [comment]
Hi @levimcclenny I want to clarify your question first. You're asking if there are any best practices for optimizing the input pipeline when using a distributed dataset? It seems you have reviewed the tf.data performance guides, but even after that your input is still a bottleneck, so you're wondering if there's anything specific you should do in the distributed dataset case. Is that correct?

Also, have you maximized the batch size to keep the GPUs busy for longer?

**Comment by levimcclenny** [comment]
Hi @nikitamaia, Actually I was having the opposite issue - my batch size was too large it would appear. Dropping it helped out pretty tremendously. I'll close the issue. 

Thanks!
